### DISCUSSION TITLE
`kedro-cmd` much faster than `kedro cmd`, should plugins adopt the former?

### DISCUSSION THREAD

**Discussion by astrojuanlu** [kedro-org/kedro#4036] [Idea]
Recently @ankatiyar has been working on improving the startup performance of Kedro commands, see discussion at https://github.com/kedro-org/kedro/issues/1476

We discarded one of the solutions, which was lazy loading commands from plugins https://github.com/kedro-org/kedro/pull/3901 because we would lose the complete list of commands in `kedro --help`.

On the other hand, @noklam's personal project, `kedro-inspect`, shows really good startup times, possibly because it's launched with `kedro-inspect` rather than `kedro inspect`. I can confirm from anecdotal evidence that it's the same for my [`kedro-init`](https://github.com/astrojuanlu/kedro-init).

It's well known that CLIs in Python are slow in general:

> Consider the example of a Python command line program (CLI) with a number of subcommands. Each subcommand may perform different tasks, requiring the import of different dependencies. But a given invocation of the program will only execute a single subcommand, or possibly none (i.e. if just --help usage info is requested). Top-level eager imports in such a program will result in the import of many modules that will never be used at all; the time spent (possibly compiling and) executing these modules is pure waste.
>
> To improve startup time, some large Python CLIs make imports lazy by manually placing imports inline into functions to delay imports of expensive subsystems. This manual approach is labor-intensive and fragile; one misplaced import or refactor can easily undo painstaking optimization work.

https://peps.python.org/pep-0690/#motivation (rejected)

And maybe the part of Kedro that collects commands from entry points can't really be optimized any further.

As a first step, should we encourage plugins to support the `kedro-cmd` version _on top of_ the currently available `kedro cmd` version?

**Comment by noklam** [comment]
First thing affected will be telemetry data
